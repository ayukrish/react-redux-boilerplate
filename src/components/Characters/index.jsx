import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Card from '../Card';
import * as styles from './characters.scss';
import Pagination from '../Pagination';

const propTypes = {
  getCharacters: PropTypes.func,
  charactersData: PropTypes.object
};

const defaultProps = {
  charactersData: {},
  getCharacters: () => {}
};

export default class Characters extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      currentPage: 1
    };
  }

  componentDidMount() {
    this.props.getCharacters(this.state.currentPage);
  }

  render() {
    const { charactersData } = this.props;
    const { results = [] } = charactersData;
    return (
      <section className={`${styles.characterWrapper} flex wrap`}>
        {results.length > 0 &&
          results.map((item) => (
            <Card
              key={item.id}
              dataObj={{
                STATUS: item && item.status,
                SPECIES: item && item.species,
                GENDER: item && item.gender,
                ORIGIN: item && item.origin && item.origin.name,
                LOCATION: item && item.location && item.location.name
              }}
              imgSrc={item && item.image}
              heading={item && item.name}
            />
          ))}
        <Pagination
          contentLength={(results && results.length) || 0}
          currentPage={this.state.currentPage}
          limit={20}
          onChange={(currentPage) => {
            this.props.getCharacters(currentPage);
            this.setState({
              currentPage
            });
          }}
        />
      </section>
    );
  }
}

Characters.propTypes = propTypes;
Characters.defaultProps = defaultProps;
