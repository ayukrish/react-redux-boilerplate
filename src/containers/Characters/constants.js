export const GET_CHARACTERS = 'GET_CHARACTERS';
export const SET_CHARACTERS = 'SET_CHARACTERS';
