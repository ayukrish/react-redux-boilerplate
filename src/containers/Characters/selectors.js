export const charactersData = (state) =>
  state.characters && state.characters.data;
