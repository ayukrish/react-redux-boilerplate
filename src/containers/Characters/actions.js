import { GET_CHARACTERS, SET_CHARACTERS } from './constants';


export const getCharacters = pageNo => ({ type: GET_CHARACTERS, data: pageNo });

export const setCharacters = data => ({ type: SET_CHARACTERS, data });
