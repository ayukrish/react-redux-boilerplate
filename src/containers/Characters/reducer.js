import { SET_CHARACTERS } from './constants';

const INITIAL_STATE = {
  data: {}
};

/**
  * Reducer - to update services state
  * @param {Object} state initial/default services state.
  */
export default function (state = INITIAL_STATE, { type, data }) {
  switch (type) {
  case SET_CHARACTERS:
    return { ...state, data };
  default:
    return state;
  }
}
