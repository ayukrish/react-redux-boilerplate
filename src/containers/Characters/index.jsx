import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getCharacters } from './actions';
import Characters from '../../components/Characters';
import { charactersData } from './selectors';

export const mapStateToProps = createStructuredSelector({
  charactersData
});

export const mapDispatchToProps = (dispatch) => ({
  getCharacters: (pageNo) => dispatch(getCharacters(pageNo))
});

export default connect(mapStateToProps, mapDispatchToProps)(Characters);
