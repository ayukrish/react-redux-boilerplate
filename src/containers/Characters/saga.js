import { takeLatest, put, call } from 'redux-saga/effects';
import { service } from '../../service';
import { enpoints } from '../../endpoint';
import { setCharacters } from './actions';
import { GET_CHARACTERS } from './constants';


export function* getCharactersSaga({ data }) {
  try {
    // data is obtained after axios call is resolved
    const response = yield call(service, {
      url: enpoints.characters,
      method: 'get',
      page: data
    });
    yield put(setCharacters(response));
  } catch (e) {
    // catch error on a bad axios call
    // alert using an alert library
  }
}

export function* getDataSaga() {
  yield takeLatest(GET_CHARACTERS, getCharactersSaga);
}
