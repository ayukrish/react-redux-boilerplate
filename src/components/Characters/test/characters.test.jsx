import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

// Components
import Characters from '../index';


describe('Characters Test Suite', () => {
  it('renders correctly without any data', () => {
    const output = shallow(
      <Characters />,
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('renders correctly without any data', () => {
    const output = shallow(
      <Characters
        getCharacters={() => { }}
        charactersData={{
          results: [{
            id: 1,
            name: 'Rick Sanchez',
            status: 'Alive',
            species: 'Human',
            gender: 'Male',
            origin: {
              name: 'Earth (C-137)',
              url: 'https://rickandmortyapi.com/api/location/1'
            },
            location: {
              name: 'Earth (Replacement Dimension)',
              url: 'https://rickandmortyapi.com/api/location/20'
            },
            image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg'
          }]
        }}
      />,
    );
    output.find('Pagination').props().onChange(1);
    expect(output.state().currentPage).toEqual(1);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
