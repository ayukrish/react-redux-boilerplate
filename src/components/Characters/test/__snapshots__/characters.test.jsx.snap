// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Test Suite renders correctly without any data 1`] = `
<section
  className="undefined flex wrap"
>
  <Pagination
    contentLength={0}
    currentPage={1}
    limit={20}
    onChange={[Function]}
  />
</section>
`;

exports[`Characters Test Suite renders correctly without any data 2`] = `
<section
  className="undefined flex wrap"
>
  <Card
    dataObj={
      Object {
        "GENDER": "Male",
        "LOCATION": "Earth (Replacement Dimension)",
        "ORIGIN": "Earth (C-137)",
        "SPECIES": "Human",
        "STATUS": "Alive",
      }
    }
    heading="Rick Sanchez"
    imgSrc="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    key="1"
  />
  <Pagination
    contentLength={1}
    currentPage={1}
    limit={20}
    onChange={[Function]}
  />
</section>
`;
